README.md
The Vector Class Library is now officially available on Github: https://github.com/vectorclass


changelog.txt
Change log for vectorclass.zip
------------------------------

Projected future version, 2018:
  * use mask registers for boolean vectors of all sizes with AVX512VL
  * use C++14
  * replace long template argument lists by constant arrays
  
2017-05-10 version 1.29
  * Reversed Apple Clang patch in version 1.28 because the problem has reoccurred in
    later versions of Clang
  
2017-05-02 version 1.28
  * Fixed problem with Apple Clang version 6.2 in vectorf128.h
  * Fixed return type for Vec8sb operator > (Vec8us, Vec8us)  
  * cpuid function modified in instrset_detect.cpp
  
2017-02-19 version 1.27
  * fixed problem with scatter functions in MS Visual Studio  
  
2016-12-21 version 1.26
  * added constant4ui template
  * fixed error for complexvec.h with clang  
  * fixed error in vectormath_exp.h for MAX_VECTOR_SIZE < 512

2016-11-25 version 1.25
  * scatter functions
  * new functions to_float for unsigned integer vectors
  * instrset_detect function can detect AVX512VL, AVX512BW, AVX512DQ
  * functions hasF16C and hasAVX512ER for detecting instruction set extensions
  * fix bugs in horizontal_and and pow(0,0) for AVX512
  * functions improved for AVX512 and AVX512VL: pow, approx_recipr,  
      approx_rsqrt
  * functions improved for AVX512DQ: 64 bit multiplication, to_double,
      32 and 64 bit rotate_left, round_to_int64, truncate_to_int64
  * functions improved for AVX512ER: approx_recipr, approx_rsqrt,
      exponential functions

2016-10-31 version 1.24
  * fix bug in Vec8uq constructor in vectori512e.h

2016-09-27 version 1.23
  * temporary fix of a problem in Clang version 3.9 inserted in vectorf128.h

2016-05-03 version 1.22
  * added optional namespace
  * fixed problem with decimal.h  

2016-04-24 version 1.21
  * fix problems with XOP option in gcc
  * improved horizontal_and/or for sse2
  * improved Vec2q and Vec4q constructor on Microsoft Visual Studio 2015
  * removed warnings by gcc option -Wcast-qual  

2015-12-04 version 1.20
  * round functions: suppress precision exception under SSE4.1 and higher
  * fix compiler problems with AVX512 multiplication in gcc version 5.1
  * fix compiler problems with pow function in Microsoft Visual Studio 2015

2015-11-14 version 1.19
  * fix various problems with Clang compiler

2015-09-25 version 1.18
  * fix compiler error for Vec8s divide_by_i(Vec8s const & x) under Clang compiler
  * fix error in Vec4d::size() in vectorf256e.h

2015-07-31 version 1.17
  * improved operator > for Vec4uq
  * more special cases in blend4q
  * nan_code functions made static inline
  * template parameter BTYPE renamed to BVTYPE in mathematical functions to avoid clash
      with macro named BTYPE in winnt.h
  * fixed bug in Vec4db constructor

2014-10-24 version 1.16
  * workaround for problem in Clang compiler extended to version 3.09 because not 
      fixed yet by Clang (vectorf128.h line 134)
  * recognize problem with Apple version of Clang reporting wrong version number
  * remove various minor problems with Clang
  * function pow(vector, int) modified to strengthen type checking and avoid compiler warnings
  * manual discusses dynamic allocation of arrays of vectors
  * various minor changes

2014-10-17 version 1.15
  * added files ranvec1.h and ranvec1.cpp for random number generator
  * constructors to make boolean vectors from their elements
  * constructors and = operators to broadcast boolean scalar into boolean vectors
  * various lookup functions improved
  * operators &, |, ^, ~, etc. defined for various boolean vectors to avoid converson
      to integer vectors
  * nmul_add functions
  * mul_add etc. moved to main header files
  * explicit fused multiply-and-add used in math functions to improve performance 
      on compilers that don't automatically insert FMA

2014-07-24 version 1.14
  * support for AVX-512f instruction set and 512-bit vectors:
      Vec16i, Vec16ui, Vec8q, Vec8uq, Vec16f, Vec8d, and corresponding boolean vectors
  * new define MAX_VECTOR_SIZE, valid values are 128, 256 and 512
  * added hyperbolic functions sinh, cosh, tanh, asinh, acosh, atanh
  * size() member function on all vector classes returns the number of elements
  * functions for conversion between boolean vectors and integer bitfields
  * extracting an element from a boolean vector now returns a bool, not an int
  * improved precision in exp2 and exp10 functions
  * various bug fixes

2014-05-11 version 1.13
  * pow function improved
  * mul_add, mul_sub, mul_sub_x functions
  * propagation of error codes through nan_code function
  * "denormal" renamed to "subnormal" everywhere, in accordance with IEEE 754-2008 standard

2014-04-20 version 1.12
  * inline implementation of mathematical functions added (vectormath_exp.h vectormath_trig.h
      vectormath_common.h)
  * vectormath.h renamed to vectormath_lib.h because a new alternative is added
  * gather functions with constant indexes
  * function sign_combine
  * function pow_const(vector, const int)
  * function pow_ratio(vector, const int, const int)
  * functions horizontal_find_first, horizontal_count
  * function recipr_sqrt removed
  * functions round_to_int64_limited, truncate_to_int64_limited, to_double_limited
  * function cubic_root renamed to cbrt
  * function atan(vector,vector) renamed to atan2
  * function if_mul
  * function Vec4i round_to_int(Vec2d)
  * operator & (float vector, boolean vector)
  * operator &= (int vector, int vector)
  * removed constructor Vec128b(int) and Vec256b(int) to avoid implicit conversion
  * removed signalling nan function
  * minor improvements in various blend and lookup functions

2014-03-01 version 1.11
  * fixed missing unsigned operators >>= in vectori256.h

2013-10-04 version 1.10
  * clear distinction between boolean vectors and integer vectors for the sake of 

vectorclass.h
/****************************  vectorclass.h   ********************************
* Author:        Agner Fog
* Date created:  2012-05-30
* Last modified: 2017-02-19
* Version:       1.27
* Project:       vector classes
* Description:
* Header file defining vector classes as interface to intrinsic functions 
* in x86 microprocessors with SSE2 and later instruction sets up to AVX512.
*
* Instructions:
* Use Gnu, Clang, Intel or Microsoft C++ compiler. Compile for the desired 
* instruction set, which must be at least SSE2. Specify the supported 
* instruction set by a command line define, e.g. __SSE4_1__ if the 
* compiler does not automatically do so.
*
* Each vector object is represented internally in the CPU as a vector
* register with 128, 256 or 512 bits.
*
* This header file includes the appropriate header files depending on the
* supported instruction set
*
* For detailed instructions, see VectorClass.pdf
*
* (c) Copyright 2012-2017 GNU General Public License www.gnu.org/licenses
******************************************************************************/
#ifndef VECTORCLASS_H
#define VECTORCLASS_H  127

// Maximum vector size, bits. Allowed values are 128, 256, 512
#ifndef MAX_VECTOR_SIZE
#define MAX_VECTOR_SIZE 256
#endif

#include "instrset.h"        // Select supported instruction set

#if INSTRSET < 2             // SSE2 required
  #error Please compile for the SSE2 instruction set or higher
#else

#include "vectori128.h"      // 128-bit integer vectors
#include "vectorf128.h"      // 128-bit floating point vectors

#if MAX_VECTOR_SIZE >= 256
#if INSTRSET >= 8
  #include "vectori256.h"    // 256-bit integer vectors, requires AVX2 instruction set
#else
  #include "vectori256e.h"   // 256-bit integer vectors, emulated
#endif  // INSTRSET >= 8
#if INSTRSET >= 7
  #include "vectorf256.h"    // 256-bit floating point vectors, requires AVX instruction set
#else
  #include "vectorf256e.h"   // 256-bit floating point vectors, emulated
#endif  //  INSTRSET >= 7
#endif  //  MAX_VECTOR_SIZE >= 256

#if MAX_VECTOR_SIZE >= 512
#if INSTRSET >= 9
  #include "vectori512.h"    // 512-bit integer vectors, requires AVX512 instruction set
  #include "vectorf512.h"    // 512-bit floating point vectors, requires AVX512 instruction set
#else
  #include "vectori512e.h"   // 512-bit integer vectors, emulated
  #include "vectorf512e.h"   // 512-bit floating point vectors, emulated
#endif  //  INSTRSET >= 9
#endif  //  MAX_VECTOR_SIZE >= 512

#endif  // INSTRSET >= 2

#endif  // VECTORCLASS_H
